src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilerAnalyzerConfigOptionsProvider.cs

```diff
@@ -8,24 +8,37 @@
 
 namespace Microsoft.CodeAnalysis.Diagnostics
 {
-    internal sealed partial class CompilerAnalyzerConfigOptionsProvider : AnalyzerConfigOptionsProvider
+    internal sealed class CompilerAnalyzerConfigOptionsProvider : AnalyzerConfigOptionsProvider
     {
         private readonly ImmutableDictionary<object, AnalyzerConfigOptions> _treeDict;
 
+        private readonly AnalyzerConfigOptions _globalOptions;
+
         public static CompilerAnalyzerConfigOptionsProvider Empty { get; }
             = new CompilerAnalyzerConfigOptionsProvider(
-                ImmutableDictionary<object, AnalyzerConfigOptions>.Empty);
+                ImmutableDictionary<object, AnalyzerConfigOptions>.Empty,
+                CompilerAnalyzerConfigOptions.Empty);
 
-        public CompilerAnalyzerConfigOptionsProvider(
-            ImmutableDictionary<object, AnalyzerConfigOptions> treeDict)
+        internal CompilerAnalyzerConfigOptionsProvider(
+            ImmutableDictionary<object, AnalyzerConfigOptions> treeDict,
+            AnalyzerConfigOptions globalOptions)
         {
             _treeDict = treeDict;
+            _globalOptions = globalOptions;
         }
 
+        public override AnalyzerConfigOptions GlobalOptions { get => _globalOptions; }
+
         public override AnalyzerConfigOptions GetOptions(SyntaxTree tree)
             => _treeDict.TryGetValue(tree, out var options) ? options : CompilerAnalyzerConfigOptions.Empty;
 
         public override AnalyzerConfigOptions GetOptions(AdditionalText textFile)
             => _treeDict.TryGetValue(textFile, out var options) ? options : CompilerAnalyzerConfigOptions.Empty;
+
+        internal CompilerAnalyzerConfigOptionsProvider WithTreeOptions(ImmutableDictionary<object, AnalyzerConfigOptions> treeDict)
+            => new CompilerAnalyzerConfigOptionsProvider(_treeDict.AddRange(treeDict), _globalOptions);
```

Comment:
It throws if the value doesn't match as well. The object we're keying off is either syntax tree or additional file though, so it should be possible to collide. And if somehow it did the options should be the same, so it would just be a noop.

---
In reply to: [437045932](https://github.com/dotnet/roslyn/pull/44875#discussion_r437045932) [](ancestors = 437045932)

Reply:
I don't think it's obvious from the API that this throws if there's a new value. The right behavior is probably to overwrite the old value with the new value. That may never happen at the current call site, but this should probably be resilient in case it's used by anyone else

Reply:
Agreed. I added `Debug.Asserts` for the current call site. Will file an issue to track post this PR.

---
In reply to: [437065445](https://github.com/dotnet/roslyn/pull/44875#discussion_r437065445) [](ancestors = 437065445)